edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Amqp/CbsNode.cs

```diff
@@ -73,33 +75,47 @@ public void RegisterLink(IAmqpLink link)
             Events.LinkRegistered(link);
         }
 
-        public async Task<bool> AuthenticateAsync(string id, Option<string> modelId)
+        public async Task<bool> AuthenticateAsync(string id, Option<string> modelId, Option<string> authChain)
         {
             try
             {
-                if (this.clientCredentialsMap.TryGetValue(id, out CredentialsInfo credentialsInfo))
+                // See if we received direct credentials for the target identity
+                ClientToken credentialsInfo;
+                bool hasCredentials = this.clientTokens.TryGetValue(id, out credentialsInfo);
+
+                // Otherwise, this could be a child Edge acting OnBehalfOf of the target,
+                // in which case we would have the actor's credentials instead
+                if (!hasCredentials)
                 {
-                    // Not -ve caching isAuthenticated here.
-                    // If incorrect credentials are sent, then it authenticates every time and fails
-                    // If correct credentials are sent later, then the authentication will succeed.
-                    if (!credentialsInfo.IsAuthenticated)
+                    Option<string> actorDeviceId = AuthChainHelpers.GetActorDeviceId(authChain);
+                    hasCredentials = actorDeviceId.Map(actor => this.clientTokens.TryGetValue(actor + $"/{CoreConstants.EdgeHubModuleId}", out credentialsInfo)).GetOrElse(false);
+                }
+
+                if (hasCredentials)
+                {
+                    if (!authenticatedClients.ContainsKey(id))
                     {
-                        using (await credentialsInfo.AsyncLock.LockAsync())
+                        IClientCredentials clientCredentials = this.clientCredentialsFactory.GetWithSasToken(
+                            credentialsInfo.DeviceId,
+                            credentialsInfo.ModuleId.OrDefault(),
+                            string.Empty,
+                            credentialsInfo.Token,
+                            true,
+                            modelId,
+                            authChain);
+
+                        if (await this.authenticator.AuthenticateAsync(clientCredentials))
                         {
-                            if (!credentialsInfo.IsAuthenticated)
-                            {
-                                credentialsInfo.ClientCredentials.ModelId = modelId;
-                                credentialsInfo.IsAuthenticated = await this.authenticator.AuthenticateAsync(credentialsInfo.ClientCredentials);
-                            }
+                            // Authentication success, add an entry for the authenticated
```

Comment:
How do we catch the token expiring?

Reply:
token expiration wasn't handled by CbsNode before, iirc the re-authenticator should catch it and have the ConnectionManager terminate the client connection if the token is expired, which should ultimately cause the CbsLink to be dropped.

---
In reply to: [489200223](https://github.com/Azure/iotedge/pull/3552#discussion_r489200223) [](ancestors = 489200223)